plugins/modules/system/xfconf.py

```diff
@@ -112,195 +111,149 @@
-            raise XFConfException('Same number of "value" and "value_type" needed')
+            raise XFConfException('Number of elements in "value" and "value_type" must be the same')
 
         # fix boolean values
-        self.value = [self._fix_bool(v[0]) if v[1] == 'bool' else v[0] for v in zip(self.value, self.value_type)]
+        self.vars.value = [fix_bool(v[0]) if v[1] == 'bool' else v[0] for v in zip(self.vars.value, value_type)]
 
         # calculates if it is an array
-        self.is_array = self.force_array or isinstance(self.previous_value, list) or values_len > 1
-        if not self.is_array:
-            self.value = self.value[0]
-            self.value_type = self.value_type[0]
-
-
-def main():
-    facts_name = "xfconf"
-    # Setup the Ansible module
-    module = AnsibleModule(
-        argument_spec=dict(
-            state=dict(default=XFConfProperty.SET,
-                       choices=XFConfProperty.VALID_STATES,
-                       type='str'),
-            channel=dict(required=True, type='str'),
-            property=dict(required=True, type='str'),
-            value_type=dict(required=False, type='list',
-                            elements='str', choices=XFConfProperty.VALID_VALUE_TYPES),
-            value=dict(required=False, type='list', elements='raw'),
-            force_array=dict(default=False, type='bool', aliases=['array']),
-        ),
-        required_if=[
-            ('state', XFConfProperty.SET, ['value', 'value_type'])
-        ],
-        supports_check_mode=True
-    )
-
-    # Force language to 'C' to ensure return values are always formatted in english, even in non-english environments
-    module.run_command_environ_update = dict(LANGUAGE='C')
-
-    state = module.params['state']
-
-    try:
-        xfconf = XFConfProperty(module)
-        xfconf.sanitize()
-
-        previous_value = xfconf.previous_value
-        facts = {
-            facts_name: dict(
-                channel=xfconf.channel,
-                property=xfconf.property,
-                value_type=xfconf.value_type,
-                value=previous_value,
-            )
-        }
-
-        if state == XFConfProperty.GET \
-                or (previous_value is not None
-                    and (state, set(previous_value)) == (XFConfProperty.SET, set(xfconf.value))):
-            module.exit_json(changed=False, ansible_facts=facts)
-            return
-
-        # If check mode, we know a change would have occurred.
-        if module.check_mode:
-            new_value = xfconf.value
+        self.vars.is_array = \
+            bool(self.module.params['force_array']) or \
+            isinstance(self.vars.previous_value, list) or \
+            values_len > 1
+
+        params = ['channel', 'property', 'create']
+        if self.vars.is_array:
+            params.append('is_array')
+        params.append('values_and_types')
+
+        extra_params = dict(values_and_types=(self.vars.value, value_type))
```

Comment:
Did you know, that `dict()` function calls are about 50% slower than dict statement `{}`?

Reply:
Even though it makes sense that placing a call in runtime will be slower than the compiler allocating a literal in compiling time, I risk saying the it is going to be negligible - I imagine we'd need hundreds of `xfconf` calls for this to bring any meaningful difference in the total time.

That said, the use of `dict()` throughout the code is driven by readability rather than performance - especially if one uses an IDE (like me with PyCharm here, btw), the colouring of `dict()` arguments is distinct from literal strings, making it easier to spot keys from values.

Do you have any references on the performance comparison with the literal `{}`? I like that kind of thing :grin:

Reply:
I found this: https://stackoverflow.com/a/6610973/7192238

I was able to replicate the exact `dis()` output in python2, but when running in python3 I got:

```
Python 3.8.5 (default, Jul 28 2020, 12:59:40) 
[GCC 9.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> def literal():
...   d = {'one': 1, 'two': 2}
... 
>>> def constructor():
...   d = dict(one=1, two=2)
... 
>>> 
>>> import dis
>>> dis.dis(literal)
  2           0 LOAD_CONST               1 (1)
              2 LOAD_CONST               2 (2)
              4 LOAD_CONST               3 (('one', 'two'))
              6 BUILD_CONST_KEY_MAP      2
              8 STORE_FAST               0 (d)
             10 LOAD_CONST               0 (None)
             12 RETURN_VALUE
>>> dis.dis(constructor)
  2           0 LOAD_GLOBAL              0 (dict)
              2 LOAD_CONST               1 (1)
              4 LOAD_CONST               2 (2)
              6 LOAD_CONST               3 (('one', 'two'))
              8 CALL_FUNCTION_KW         2
             10 STORE_FAST               0 (d)
             12 LOAD_CONST               0 (None)
             14 RETURN_VALUE
>>> 
```
But it still makes quite a difference:
```
$ python3 -m timeit "d = dict(a=1, b=2, c=3, d=4, e=5)"                  
1000000 loops, best of 5: 339 nsec per loop
$ python3 -m timeit "d = {'a':1, 'b':2, 'c':3, 'd':4, 'e':5}"                                             
2000000 loops, best of 5: 132 nsec per loop
```

Reply:
I just thought you may find this fun fact interesting because in my experience a lot of people are not aware of that.  
It was also very late and I was writing from my mobile, that's why I was unable to post any reference links, but I'm glad you did your own reserach :slightly_smiling_face: 

> That said, the use of `dict()` throughout the code is driven by readability rather than performance - especially if one uses an IDE (like me with PyCharm here, btw), the colouring of dict() arguments is distinct from literal strings, making it easier to spot keys from values.

Interesting! I'm also using Pycharm on daily basis, but I never paid any attention to it.  
On the other hand, JetBrains IDEs are highly customizable, so I bet there is a way to achieve the same behavior for literal `{}` :wink:

Reply:
>  I like that kind of thing :grin: 

So maybe you will find the [`collections.deque`](https://docs.python.org/3/library/collections.html#collections.deque) also interesting.  
TL;DR The implementation of `list` in Python C code is very memory-inefficient, when it comes to adding/subtracting list elements.